File: 3351-maximize-happiness-of-selected-children/maximize-happiness-of-selected-children.cpp
class Solution {
public:
    long long maximumHappinessSum(vector<int>& happiness, int k) {
        priority_queue<int> maxHeap;
        for(auto it : happiness) {
            maxHeap.push(it);
        }
        long long maxHappiness = 0;
        int turns = 0;
        while(turns < k) {
            maxHappiness += max((maxHeap.top() - turns), 0);
            maxHeap.pop();
            turns ++;
        }

        return maxHappiness; 
    }
};

File: 3351-maximize-happiness-of-selected-children/maximize-happiness-of-selected-children.java
class Solution {
    public long maximumHappinessSum(int[] happiness, int k) {
        Queue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        for(int i =0 ; i < happiness.length; i++){
            maxHeap.add(happiness[i]);
        }
        long res = 0;
        int turns = 0;
        while(k-- > 0) {
            int cur = maxHeap.poll() - turns;
            if(cur <= 0) {
                return res;
            }
            res += cur;
            turns++;
        }

        return res;
    }
}